Build hash table while adding input rows for left semi and anti join

liujiayi771 opened the pull request:
As discussed in the [issue](https://github.com/facebookincubator/velox/issues/6577), during the hash build process, left semi join and anti join can deduplicate the input rows based on the join key. However, Velox's hash build addInput process adds all inputs to the RowContainer, which can result in significant memory wastage in certain scenarios, such as TPCDS Q14 and Q95. To address this, we can construct the hash table directly during data input and utilize the existing allowDuplicates parameter of the hashTable to remove duplicate data without storing it in the RowContainer. This process is similar to constructing a hash table in the hash aggregation process.

Due to Velox's hash build potentially having multiple drivers executing, in this scenario, duplicate data can only be removed for individual driver inputs. However, in the case of single-driver execution mode, it is possible to remove all duplicate data.

liujiayi771 commented:
I use Gluten to perform testing, and I have observed significant reductions in peak memory usage and execution time for hash build in the case of a 1TB TPCDS Q95 dataset.
create hash table after all inputs have been added to RowContainer:
![image](https://github.com/facebookincubator/velox/assets/13622031/3a986d86-3fbf-4e34-882f-0725cfed55b0)

stream input rows to hash table, and deduplicate the input rows:
![image](https://github.com/facebookincubator/velox/assets/13622031/ea6329c9-9c9e-4640-a690-ff391c36b0c2)

liujiayi771 commented:
Hi @mbasmanova. Do you have any suggestions for this optimization?

mbasmanova commented on velox/exec/HashTable.h:
```diff
@@ -751,7 +751,7 @@ class HashTable : public BaseHashTable {
   // join build sides.
   int32_t reservePct() const {
-    return isJoinBuild_ ? 0 : 50;
+    return (isJoinBuild_ && rows_->nextOffset()) ? 0 : 50;
```
Is this an unrelated fix? Maybe extract it into a separate PR and explain what it does.

liujiayi771 replied:
Currently, there is no variable in the HashTable that can represent whether duplicate data is allowed. However, rows_->nextOffset() can be used to indicate that duplicate data is not allowed. For example, in the HashTable of a left semi join, rows_->nextOffset() is 0. Introducing a new variable `allowDuplicates` in HashTable here may make it clearer.

I feel that if HashBuild builds a hash table while adding input rows, it should also reserve some values for new keys, just like agg. I'm not sure if my understanding of `reservePct` is correct, please correct me if I'm wrong.

mbasmanova reviewed (commented):
@liujiayi771 Thank you for the optimization. Overall this makes sense to me. Just to clarify, my understand is the following.

Before this change, HashBuild operator first added all input data into RowContainer, then built a hash table over it. Now, HashBuild operator is building hash table while adding input rows. In case when HashBuild operator runs single-threaded there is no need to build hash table again after processing all input. In case when HashBuild operator runs multi-threaded it is still necessary to re-build the hash table after all input has been received an combined.

liujiayi771 commented:
@mbasmanova Yes. For single-threaded scenarios, further optimization can be done to avoid rebuilding the hash table.

liujiayi771 commented:
Hi @mbasmanova. During the process of fixing the UT, I discovered that regardless of whether it is a single-threaded hash build or not, it is necessary to call `table_->prepareJoinTabl`e in the `finishHashBuild` to perform a rehash. Although the hash table is already correct and usable at this point, there may still be some rows that have not been added to `uniqueValues_` in `VectorHasher`. This may result in the generated dynamic filter not including all the values. Therefore, during the final rehash, it will execute `table->analyze` to add all rows to uniqueValues_.

liujiayi771 commented:
Hi @xiaoxmeng.
Currently, the tests related to the "join builder" in SharedArbitrationTest are failing. I'm not very familiar with arbitration, could you give me some advice? Can the current modifications to the hash table build process be compatible with the current arbitration for HashBuild?

In addition, I have identified some minor points. For example, the current anti join deduplicates and does not store non-join key columns, which reduces memory usage. Therefore, we may need to add a seed to ensure that `newVector` in  SharedArbitrationTest generates a different vector each time it is called. 

Additionally, it may be necessary to reduce the `joinMemoryUsage` in the tests to trigger the execution of fakeAllocation. However, currently SharedArbitrationTest.reclaimFromJoinBuilder still cannot pass.

liujiayi771 commented:
cc @Yohahaha

liujiayi771 commented:
@xiaoxmeng Could you help review this change?

xiaodouchen commented:
@liujiayi771 @mbasmanova  Is there any latest progress on this issue?  I am very interested in it. 

I've been exploring the implementation details and have a question regarding the efficiency of the hash table construction within the addInput method. In cases where the input data has a very low duplication rate, it will be inefficient to build the hash table at this stage given that it will finally re-build the hash table.

Additionally, for left semi join and anti join, is there an opportunity to only optimize by storing only the join keys in the RowContainer?

mbasmanova commented:
@xiaodouchen 

> or left semi join and anti join, is there an opportunity to only optimize by storing only the join keys in the RowContainer?

I believe this should be done in the optimizer. The query plan passed to Velox should not have non-join keys on the build side.

mbasmanova commented:
> In cases where the input data has a very low duplication rate, it will be inefficient to build the hash table at this stage given that it will finally re-build the hash table.

@xiaodouchen That's a good point. Perhaps, this logic can be adaptive. If after processing some data it sees low duplication rate, it stops building a hash table.

liujiayi771 commented:
@xiaodouchen @mbasmanova If there are a lot of duplicate data, it will indeed be time-consuming when rehashing at the end. A similar approach to the early abandonment in 'agg' can be adopted. 

Regarding not storing non-join keys in the row container, I understand that this has already been implemented. You can see it here [https://github.com/facebookincubator/velox/pull/7066/files#diff-590ded45d84bc64a2cb5d135394bf677852fdcbb4bdc1e99fc2c7746153d9f4bR166], but the actual situation is that these columns will not be passed down. The optimizer will pass only the join key columns through a project.

liujiayi771 commented:
@xiaodouchen The main purpose of this modification is to reduce memory consumption caused by duplicate data, and the performance improvement also primarily comes from the reduction in data volume. I have tested 10T TPCDS q14 and q95 under Spark + Velox and there is a very significant performance improvement. Memory usage has dropped by more than a hundredfold for join operator, and execution speed has increased by more than 20%.

In the Velox's own TPCH benchmark, due to the smaller data volume, the improvement in speed may not be as noticeable. I can run a TPCH test on my cluster using Spark in the next couple of days to see the memory usage and speed improvement of the semi join, and I will reply with the results afterwards. I will look into optimizations based on data duplication rates in the next two weeks.

liujiayi771 commented:
Hi @xiaodouchen, I conducted a simple test on TPCH 1TB q20, q21, and q22. The optimizations made in this PR did not have a significant impact on these three queries. This could be due to the duplication rate of the build-side data. There was not much improvement in terms of execution time for these queries; the time remained mostly the same. The memory usage for q20 and q21 was also similar, with only q22 showing a noticeable decrease in memory usage for left semi-join.

zhli1142015 commented:
Any progress for this PR?
Thanks.

liujiayi771 commented:
> Any progress for this PR? Thanks.
@zhli1142015 We've been using this PR internally, and the results have been quite positive. It's been a while since this PR was last updated, but I can continue to develop the dynamic enabling based on the duplication rate. Have you tested this pull request? Do you have any suggestions?

zhli1142015 commented:
> > Any progress for this PR? Thanks.
> 
> @zhli1142015 We've been using this PR internally, and the results have been quite positive. It's been a while since this PR was last updated, but I can continue to develop the dynamic enabling based on the duplication rate. Have you tested this pull request? Do you have any suggestions?

Thanks for your update. We observed a silimar issue, we want to test this internally also. @ayushi-agarwal

Thanks

liujiayi771 commented:
Hi @mbasmanova @xiaoxmeng @Yuhta. I have recently spent some time updating this PR to include support for falling back to the original mode (only adding rows to the row container in `addInput`) when the duplicate ratio is low. I have also adapted some logic for join spill. 

This PR addresses the issue of not needing to save duplicate data to the row container for left semi and anti join when the join key has a high duplicate rate in the build side input. Storing these duplicate data can consume a significant amount of memory, especially in TPCDS Q95. Due to the current design of Velox, which first adds input and then builds the hash table, it is not possible to deduplicate in the `addInput` phase. The current PR supports directly using the same mode as building the hash table in the agg, meaning building the hash table and deduplicating upon input, so the data in the hash table row container is deduplicated.

I think that Velox's join could benefit from this optimization, and I wonder if others feel that the current implementation has any issues. Alternatively, we could consider retaining the original mode but implementing a method to remove duplicate data from the row container and free up memory after `addInput` is completed, during the hash table build phase.

mbasmanova commented:
@liujiayi771 

Thank you for working on this. Overall this optimization sounds useful. I particularly like that it seems you have made it adaptive, i.e. stop de-duplication if there are not too many duplicates.

I imagine this logic is non-trivial. It would be nice to update the PR description to provide a bit more context on the overall design and any tricky parts that you needed to handle. In particular, it would be nice to describe how the adaptivity works and which configs (if any) are introduced to control it.

Finally, wondering how did you make sure the changes are working 100%. What testing strategy did you use? How are you making sure test coverage is "complete"?

liujiayi771 commented:
Hi @mbasmanova.
I will take some time to organize the design of this PR, and I will create an issue to discuss this design.

> how did you make sure the changes are working 100%. What testing strategy did you use? How are you making sure test coverage is "complete"?

Are you referring to the adaptive stop de-duplicate? I have added two test cases, with `kAbandonBuildNoDupHashMinPct` and `kAbandonBuildNoDupHashMinRows` set to relatively low values to make it abandon de-duplicates. However, these tests may not be comprehensive enough; it might be better to add this test in the join fuzzer.

For the additional testing of this optimization, I am mainly relying on the join fuzzer and the existing test cases for left semi and anti-join to test this optimization.

mbasmanova commented:
@liujiayi771 

> I will take some time to organize the design of this PR, and I will create an issue to discuss this design.

That sounds great. Thank you.

> However, these tests may not be comprehensive enough; it might be better to add this test in the join fuzzer.

Once the design is ready, let's review to figure out how to extend join fuzzer to provide good coverage.

aditi-pandit commented:
@liujiayi771 : Are you planning to continue this optimization ? This might be helpful for TPC-DS.

liujiayi771 commented:
@aditi-pandit This patch has already been in use in our internal version and it has shown excellent results for q95. Could you help to review? I would like to continue developing this optimization.

aditi-pandit commented:
@liujiayi771 : We (IBM) are interested as well. Can you rebase your code ? Will help me to pull this internally and test on our side. 

Also, please can you provide your design in a doc or issue. I can help you with the reviews.

XinShuoWang commented on velox/exec/HashTable.cpp:
```diff
@@ -1731,8 +1734,21 @@ template <bool ignoreNullKeys>
+    if (table_ != nullptr) {
+      // Set table_ to nullptr to trigger rehash.
+      rows_->pool()->freeContiguous(tableAllocation_);
+      table_ = nullptr;
```
Is it necessary to clean up `capacity_ ` and `hashMode_` variables?

liujiayi771 replied:
No need to clean up other variables. Setting `table_ = nullptr` here is to allow entering the rehash process in the checkSize method, and it is not intended for data cleanup.

XinShuoWang replied:
```cpp
TEST_F(HashJoinTest, semiJoinDeduplicateResetCapacity) {
  // The initial size of HashTable is 2048. After detecting the
  // kAbandonBuildNoDupHashMinRows row, it is found that the duplication rate is
  // lower than kAbandonBuildNoDupHashMinPct, so no more data is inserted into
  // HashTable (the capacity_ variable does not change afterwards). However,
  // data will continue to be added to RowContainer, causing numDistinct_ to be
  // greater than 2048, and eventually an error is reported in the
  // HashTable<ignoreNullKeys>::checkSize function.
  const int vectorSize = 10, batches = 210;
  auto probeVectors = makeBatches(batches, [&](int32_t /*unused*/) {
    return makeRowVector({
        // Join Key is double -> VectorHasher::typeKindSupportsValueIds will
        // return false -> HashMode is kHash
        makeFlatVector<double>(
            vectorSize, [&](vector_size_t /*row*/) { return rand(); }),
        makeFlatVector<int64_t>(
            vectorSize, [&](vector_size_t /*row*/) { return rand(); }),
    });
  });

  auto buildVectors = makeBatches(batches, [&](int32_t batch) {
    return makeRowVector({
        makeFlatVector<double>(
            vectorSize, [&](vector_size_t /*row*/) { return rand(); }),
        makeFlatVector<int64_t>(
            vectorSize, [&](vector_size_t /*row*/) { return rand(); }),
    });
  });

  createDuckDbTable("t", probeVectors);
  createDuckDbTable("u", buildVectors);

  auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
  auto plan = PlanBuilder(planNodeIdGenerator)
                  .values(probeVectors)
                  .project({"c0 AS t0", "c1 AS t1"})
                  .hashJoin(
                      {"t0"},
                      {"u0"},
                      PlanBuilder(planNodeIdGenerator)
                          .values(buildVectors)
                          .project({"c0 AS u0", "c1 AS u1"})
                          .planNode(),
                      "",
                      {"t0", "t1", "match"},
                      core::JoinType::kLeftSemiProject)
                  .planNode();

  HashJoinBuilder(*pool_, duckDbQueryRunner_, driverExecutor_.get())
      .config(core::QueryConfig::kAbandonBuildNoDupHashMinRows, "10")
      .config(core::QueryConfig::kAbandonBuildNoDupHashMinPct, "50")
      .numDrivers(1)
      .checkSpillStats(false)
      .planNode(plan)
      .referenceQuery(
          "SELECT t.c0, t.c1, EXISTS (SELECT * FROM u WHERE t.c0 = u.c0) FROM t")
      .run();
}
```
This case will not work without resetting the capacity variable.

liujiayi771 replied:
@XinShuoWang Thank you for your review. I understand that when "abandon" is triggered, the `capacity_` is not accurate and needs to be reset. I will make the necessary modifications and incorporate your test cases.

aditi-pandit commented:
@liujiayi771 : Thanks for this code. Would you be able to write a micro-benchmark that we can measure and review ? It would help us tune the config parameters as well.

liujiayi771 commented:
@aditi-pandit Sure, I will add some benchmarks related to performance and memory usage.

czentgr commented:
@liujiayi771 FYI, there are test failures in the Ubuntu build in the HashJoinTest test. So this could indicate problems?

E.g.
```
 [ RUN      ] HashJoinTest.joinBuildSpillError
/home/runner/work/velox/velox/velox/velox/exec/tests/HashJoinTest.cpp:7659: Failure
Failed
Expected an exception
...
```

czentgr commented:
@liujiayi771 Thank you!. Can you please also rebase and squash the commits and a single commit helps tracking it.

aditi-pandit commented:
@liujiayi771 : Please can you also add documentation of the properties in https://github.com/facebookincubator/velox/blob/main/velox/docs/configs.rst

liujiayi771 commented:
> @liujiayi771 : Please can you also add documentation of the properties in https://github.com/facebookincubator/velox/blob/main/velox/docs/configs.rst

Sure, I've been quite busy recently. I expect to complete the benchmark and other modifications in about two weeks.

liujiayi771 commented:
> @liujiayi771 FYI, there are test failures in the Ubuntu build in the HashJoinTest test. So this could indicate problems?

I was able to reproduce the issue you mentioned in the Ubuntu Docker env. I suspect that it is related to the fact that after I modified the left semi/anti join, the memory consumption decreased, which led to the triggering of spills being different from before, resulting in the test results not meeting expectations. I will try to fix the test case.

liujiayi771 commented:
@czentgr I was able to reproduce the issue you mentioned in the Ubuntu Docker env. I suspect that it is related to the fact that after I modified the left semi/anti join, the memory consumption decreased, which led to the triggering of spills being different from before, resulting in the test results not meeting expectations.
For joinBuildSpillError test case, I changed kMemoryCapacity to 27 << 20, and it passed.

aditi-pandit commented:
@xiaoxmeng , @Yuhta : Please can you take a look at this code. This is very helpful for TPC-DS queries.

aditi-pandit reviewed (commented):
@liujiayi771 : Thanks for this code. It is shaping well. Do you have a sample benchmark output to share ?

liujiayi771 commented:
> @liujiayi771 : Thanks for this code. It is shaping well. Do you have a sample benchmark output to share ?

You can check the benchmark output in https://github.com/facebookincubator/velox/issues/11212.

czentgr commented:
The HashJoin test failure reproduced locally:

```
[ RUN      ] HashJoinTest.dynamicFilters
I0211 09:58:10.263239 1319148 Compression.cpp:629] Initialized zstd compressor with compression level 7
/root/lakehouse/velox/velox/exec/tests/HashJoinTest.cpp:4682: Failure
Expected: (getInputPositions(task, 1)) < (numRowsProbe * numSplits), actual: 3330 vs 3330
Google Test trace:
/root/lakehouse/velox/velox/exec/tests/HashJoinTest.cpp:4669: hasSpill:false
/root/lakehouse/velox/velox/exec/tests/HashJoinTest.cpp:665: Without Spill
/root/lakehouse/velox/velox/exec/tests/utils/QueryAssertions.cpp:1182: Failure
Failed
Expected 1000, got 3330
2330 extra rows, 0 missing rows
10 of extra rows:
        1
        1
        1
        1
        1
        1
        1
        1
        1
        1

0 of missing rows:

Note: DuckDB only supports timestamps of millisecond precision. If this test involves timestamp inputs, please make sure you use the right precision.
DuckDB query: SELECT t.c1 + 1 FROM t, u WHERE t.c0 = u.c0 AND t.c0 < 500
Google Test trace:
/root/lakehouse/velox/velox/exec/tests/HashJoinTest.cpp:665: Without Spill
/root/lakehouse/velox/velox/exec/tests/HashJoinTest.cpp:4805: Failure
Expected: (getInputPositions(task, 1)) < (numRowsProbe * numSplits), actual: 3330 vs 3330
Google Test trace:
/root/lakehouse/velox/velox/exec/tests/HashJoinTest.cpp:4792: hasSpill:false
/root/lakehouse/velox/velox/exec/tests/HashJoinTest.cpp:665: Without Spill
/root/lakehouse/velox/velox/exec/tests/utils/QueryAssertions.cpp:1182: Failure
Failed
Expected 830, got 2450
1620 extra rows, 0 missing rows
10 of extra rows:
        1
        1
        1
        1
        1
        1
        1
        1
        1
        1

0 of missing rows:

Note: DuckDB only supports timestamps of millisecond precision. If this test involves timestamp inputs, please make sure you use the right precision.
DuckDB query: SELECT t.c1 + 1 FROM t, u WHERE t.c0 = u.c0 AND t.c0 < 200
Google Test trace:
/root/lakehouse/velox/velox/exec/tests/HashJoinTest.cpp:665: Without Spill
/root/lakehouse/velox/velox/exec/tests/utils/QueryAssertions.cpp:1182: Failure
Failed
Expected 830, got 2450
1620 extra rows, 0 missing rows
10 of extra rows:
        1
        1
        1
        1
        1
        1
        1
        1
        1
        1

0 of missing rows:

Note: DuckDB only supports timestamps of millisecond precision. If this test involves timestamp inputs, please make sure you use the right precision.
DuckDB query: SELECT t.c1 + 1 FROM t WHERE t.c0 IN (SELECT c0 FROM u) AND t.c0 < 200
Google Test trace:
/root/lakehouse/velox/velox/exec/tests/HashJoinTest.cpp:665: Without Spill
[  FAILED  ] HashJoinTest.dynamicFilters (4705 ms)
```

aditi-pandit commented:
@liujiayi771 : Thanks for your code contribution and it has been looking good to me. We also have been using it internally for a long time now. Would prefer Meta folks to give rest of go ahead.

Can you help rebase your code for further review.

jinchengchenghh commented on velox/exec/HashBuild.cpp:
```diff
@@ -1203,4 +1235,10 @@ void HashBuildSpiller::extractSpill(
+bool HashBuild::abandonBuildNoDupHashEarly(int64_t numDistinct) const {
+  VELOX_CHECK(dropDuplicates_);
+  return numHashInputRows_ > abandonBuildNoDupHashMinRows_ &&
+      100 * numDistinct / numHashInputRows_ >= abandonBuildNoDupHashMinPct_;
```
numDistinct / numHashInputRows_ >=  abandonBuildNoDupHashMinPct_ / 100 for overflow case

jinchengchenghh commented on velox/exec/HashBuild.h:
```diff
@@ -265,6 +271,11 @@ class HashBuild final : public Operator {
 
+  // Indicates whether drop duplicate rows. Rows containing duplicate keys
+  // can be removed for left semi and anti join.
+  bool dropDuplicates_{false};
```
const

jinchengchenghh commented on velox/exec/HashTable.cpp:
```diff
@@ -57,6 +57,7 @@ HashTable<ignoreNullKeys>::HashTable(
       pool_(pool),
       minTableSizeForParallelJoinBuild_(minTableSizeForParallelJoinBuild),
       isJoinBuild_(isJoinBuild),
+      joinBuildNoDuplicates_(!allowDuplicates),
```
joinBuildNoDuplicates_ -> joinBuildDistinct

liujiayi771 replied:
Do you think I need to change all instances of "NoDup" to "Distinct"? There are several variables that use "NoDup," including abandonBuildNoDupHashMinRows_, abandonBuildNoDupHashMinPct_, and others. The reason I use "NoDup" is that HashTable has the concept of "drop duplicates".

jinchengchenghh replied:
Sounds reasonable

jinchengchenghh reviewed (commented):
Thanks

liujiayi771 commented on velox/exec/benchmarks/HashJoinStreamBuildBenchmark.cpp:
```diff
@@ -0,0 +1,389 @@
+  folly::runBenchmarks();
+
+  for (auto& result : results) {
+    std::cout << result.toString() << std::endl;
```
The main purpose here is to print out `peakMemoryBytes`, `AbandonPct`, `isAbandon` to demonstrate the optimization effect, but Folly Benchmark cannot output this information.
```
DupFactor:2,  AbandonPct:90, HashMode:NORMALIZED_KEY, isAbandon:false, totalClock:60448ms, peakMemoryBytes:34.25MB
```

jinchengchenghh commented:
In the multiple drivers executing scenario, may it be possible regression?

liujiayi771 commented:
> In the multiple drivers executing scenario, may it be possible regression?

In the case of multiple drivers, complete deduplication cannot be achieved. However, when the data duplication rate is relatively high, there are still benefits. Moreover, because of the fallback mechanism, there won't be significant performance degradation.

jinchengchenghh commented:
LGTM. Do we need to set the config to make this feature disabled default? we can set it to reasonable value when enable it. This may help the feature merge.

liujiayi771 commented:
> But still didn't see any diff in the peak memory and perf. 

@FelixYBW 
It is disabled by default. You need to set `kAbandonBuildNoDupHashMinPct` to a value greater than 0, for example, 60. Have you configured it?

FelixYBW commented:
> kAbandonBuildNoDupHashMinPct

No, 

> kAbandonBuildNoDupHashMinPct

Do I only need to set kAbandonBuildNoDupHashMinPct=60? It's odd I tried but still no difference.

liujiayi771 commented:
@FelixYBW I have tested this PR for scale=100G, it will not work when kAbandonBuildNoDupHashMinPct=60. It might be the distinct join keys occupy a relatively large proportion in the first 100,000 rows, which triggers the abandonment strategy. When I set kAbandonBuildNoDupHashMinPct to 100, it no longer triggers the abandonment, and the peak memory usage dropped significantly.
<img width="500" height="311" alt="image" src="https://github.com/user-attachments/assets/c34c1ad3-f91b-4e0f-a579-ecff4729ef24" />

liujiayi771 commented:
@FelixYBW You can check the benchmark results in the #6577. If the data duplication rate is not high, this optimization could have a negative effect. Therefore, we've added a feature to disable the optimization based on the duplication rate. For TPCDS q95, the join key has a very high duplication rate, so the improvement is significant.

FelixYBW commented:
> @FelixYBW You can check the benchmark results in the #11212. If the data duplication rate is not high, this optimization could have a negative effect. Therefore, we've added a feature to disable the optimization based on the duplication rate. For TPCDS q95, the join key has a very high duplication rate, so the improvement is significant.

Thank you, Could you create a PR in Gluten to make it configurable? Or do we already have?

FelixYBW commented:
@pedroerp We have confirmed the PR can benefit TPCDS Q95 from both memory consumption and elapsed time. Can you help to review and merge?

pedroerp commented:
Perhaps @xiaoxmeng or @tanjialiang could take a look as they have more context on this code.

jinchengchenghh commented:
Could you help rebase to fix the conflict? @liujiayi771

FelixYBW commented:
> Could you help rebase to fix the conflict? @liujiayi771

@liujiayi771 @zhouyuan helped to resolved conflicts here, in case it helps.

https://github.com/facebookincubator/velox/compare/main...IBM:velox:oap_7066

tanjialiang commented on velox/exec/HashBuild.cpp:
```diff
@@ -1240,4 +1277,10 @@ void HashBuildSpiller::extractSpill(
   }
 }
+
+bool HashBuild::abandonBuildNoDupHashEarly(int64_t numDistinct) const {
```
Curious why do we abandon? Why not we always do the deduped way?

liujiayi771 replied:
@tanjialiang You can check the benchmark tests in #11212. When the duplicate rate of join keys in a left semi/anti join is not high, deduplication incurs additional overhead. Additionally, perhaps making this optimization disabled by default would make it more acceptable for the community to merge?

tanjialiang commented on velox/exec/HashTable.cpp:
```diff
@@ -57,6 +57,7 @@ HashTable<ignoreNullKeys>::HashTable(
       pool_(pool),
       minTableSizeForParallelJoinBuild_(minTableSizeForParallelJoinBuild),
       isJoinBuild_(isJoinBuild),
+      joinBuildNoDuplicates_(!allowDuplicates),
```
Why don't we just reuse the name allowDuplicates_ which is more straightforward? Combining with isJoinBuild_ they can tell the use case scenario

liujiayi771 replied:
Yes, we can reuse it by using `!allowDuplicates` to represent `noDuplicates`.

tanjialiang commented on velox/exec/HashBuild.cpp:
```diff
@@ -377,6 +387,31 @@ void HashBuild::addInput(RowVectorPtr input) {
   }
 
+  if (dropDuplicates_ && !abandonBuildNoDupHash_) {
+    const bool abandonEarly = abandonBuildNoDupHashEarly(table_->numDistinct());
```
Curious why we abandon? In what scenarios the dedup way is less memory/cpu efficient? If for all general cases it is more efficient, should we always make this special case (dedup leftsemi and anti) without abandoning?

tanjialiang commented on velox/exec/HashTable.cpp:
```diff
@@ -1709,8 +1712,20 @@ template <bool ignoreNullKeys>
+    folly::Executor* executor,
+    bool dropDuplicates) {
   buildExecutor_ = executor;
+  if (dropDuplicates) {
```
At this point the dedup could have possibly been abandoned?

liujiayi771 replied:
Yes, but we still need to clean up `table_`, because as long as HashBuild has ever entered the deduplication process, we need to perform rehash.

tanjialiang commented on velox/exec/HashTable.cpp:
```diff
@@ -1740,6 +1755,11 @@ void HashTable<ignoreNullKeys>::prepareJoinTable(
     }
     if (useValueIds) {
       for (auto& other : otherTables_) {
+        if (dropDuplicates) {
```
ditto

tanjialiang commented on velox/exec/HashBuild.cpp:
```diff
@@ -1240,4 +1277,10 @@ void HashBuildSpiller::extractSpill(
+bool HashBuild::abandonBuildNoDupHashEarly(int64_t numDistinct) const {
+  VELOX_CHECK(dropDuplicates_);
+  return numHashInputRows_ > abandonBuildNoDupHashMinRows_ &&
+      numDistinct / numHashInputRows_ >= abandonBuildNoDupHashMinPct_ / 100;
```
numDistinct (int) / numHashInputRows_ (int) will always result in 0 if numDistinct < numHashInputRows?
abandonBuildNoDupHashMinPct_ (int) / 100 will always result in 0 if abandonBuildNoDupHashMinPct_ < 100?
So the second condition will always hold true?

tanjialiang replied:
Additionally I am wondering why we need two configs for this feature? In your benchmark (which I think might need to be redone) the size is fixed. Did you observe num rows being a significant factor impacting performance when it comes to abandon threshold? The reason I'm asking is this config is hard for users to set and tune. Based on your experiments do you think it will be reasonable to drop the num row threshold config?

liujiayi771 replied:
> Based on your experiments do you think it will be reasonable to drop the num row threshold config?
Here, it's the same as the partial aggregation abandonment. We need a sufficient sample size to determine the duplication rate, and the higher the threshold, the more accurate the duplication rate assessment becomes.

In most cases, users don't need to adjust this threshold config. However, for scenarios with a large number of input rows where users have some understanding of the data's duplication pattern, they can fine-tune the triggering point for the abandon decision by adjusting the row count threshold.

Without this parameter, we would still need a fixed-size threshold to trigger duplication rate check.

tanjialiang commented on velox/exec/HashTable.h:
```diff
@@ -310,7 +310,19 @@ class BaseHashTable {
+  /// table. This function is used to indicate that the join hash table will not
+  /// be built during the addInput phase, and the input data will also not be
+  /// deduplicated.
+  virtual void joinTableMayHaveDuplicates() = 0;
```
prefer regular setter
setAllowDuplicates(bool)

tanjialiang commented on velox/exec/HashTable.h:
```diff
@@ -587,6 +599,10 @@ class HashTable : public BaseHashTable {
     return hasDuplicates_.check();
   }
 
+  void joinTableMayHaveDuplicates() override {
```
Could we make the names match?

tanjialiang commented on velox/exec/HashTable.h:
```diff
@@ -310,7 +310,19 @@ class BaseHashTable {
+  /// like other join types by reading all the data before building the hash
+  /// table. This function is used to indicate that the join hash table will not
+  /// be built during the addInput phase, and the input data will also not be
+  /// deduplicated.
```
Can we also indicate this does not impact the containing row container in any means?

tanjialiang commented on velox/core/QueryConfig.h:
```diff
@@ -181,6 +181,12 @@ class QueryConfig {
   static constexpr const char* kAbandonPartialTopNRowNumberMinPct =
       "abandon_partial_topn_row_number_min_pct";
 
+  static constexpr const char* kAbandonBuildNoDupHashMinRows =
```
kAbandonHashBuildDedupMinRows as there are other join types as well

tanjialiang commented on velox/core/QueryConfig.h:
```diff
@@ -181,6 +181,12 @@ class QueryConfig {
+  static constexpr const char* kAbandonBuildNoDupHashMinRows =
+      "abandon_build_no_dup_hash_min_rows";
+
+  static constexpr const char* kAbandonBuildNoDupHashMinPct =
```
ditto

FelixYBW commented:
@tanjialiang any more comments?

tanjialiang commented:
> @tanjialiatan Could you help review again?



> @tanjialiang any more comments?

Hey sorry for the delay, I'll let @xiaoxmeng do to a final pass.

zhouyuan commented on velox/docs/configs.rst:
```diff
@@ -43,6 +43,16 @@ Generic Configuration
+     - 100,000
+     - Number of input rows to receive before starting to check whether to abandon building a HashTable without
+       duplicates in HashBuild for left semi/anti join.
+   * - abandon_build_no_dup_hash_min_pct
```
Just a nit: the current config name is a bit confusing. How about 'dedup hashmap' to better reflect the underlying change for this feature?

liujiayi771 replied:
Yes, I will change it to `abandon_dedup_hashmap_min_pct`

tanjialiang commented:
Hi @liujiayi771 , I have communicated with @xiaoxmeng . He will do a final pass and we are going to be good.

xiaoxmeng commented on velox/exec/benchmarks/HashJoinStreamBuildBenchmark.cpp:
```diff
@@ -0,0 +1,393 @@
+  }
+};
+
+class HashJoinStreamBuildBenchmark : public VectorTestBase {
```
s/HashJoinStreamBuildBenchmark/HashJoinBuildBenchmark/? Thanks!

xiaoxmeng commented on velox/exec/HashTable.cpp:
```diff
@@ -1716,8 +1719,20 @@ template <bool ignoreNullKeys>
     int8_t spillInputStartPartitionBit,
-    folly::Executor* executor) {
+    folly::Executor* executor,
+    bool dropDuplicates) {
```
can we put executor such env parameter last?

xiaoxmeng commented on velox/exec/HashBuild.cpp:
```diff
@@ -377,6 +387,31 @@ void HashBuild::addInput(RowVectorPtr input) {
+      // that was previously inserted into the hash table is already in the
+      // RowContainer.
+      addRuntimeStat("abandonBuildNoDupHash", RuntimeCounter(1));
+      abandonHashBuildDedup_ = true;
```
Put this into a function abandonHashBuildDedup()?

xiaoxmeng commented on velox/exec/HashBuild.cpp:
```diff
@@ -170,15 +173,22 @@ void HashBuild::setupTable() {
           pool());
     }
   }
+  lookup_ = std::make_unique<HashLookup>(table_->hashers(), pool());
```
Shall we only create if dedup is enabled. And also add to reset it in abandon function as commented below? Thanks!
